hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableGenericRecordBuilder.java

```diff
@@ -0,0 +1,187 @@
+    }
+
+    @Override
+    public GenericRecord.Builder writeDouble(String fieldName, double value) {
+        return write(fieldName, value, FieldType.DOUBLE);
+    }
+
+    @Override
+    public GenericRecord.Builder writeFloat(String fieldName, float value) {
+        return write(fieldName, value, FieldType.FLOAT);
+    }
+
+    @Override
+    public GenericRecord.Builder writeShort(String fieldName, short value) {
+        return write(fieldName, value, FieldType.SHORT);
+    }
+
+    @Override
+    public GenericRecord.Builder writeGenericRecord(String fieldName, @Nullable GenericRecord value) {
+        return write(fieldName, value, FieldType.PORTABLE);
+    }
+
+    @Override
+    public GenericRecord.Builder writeGenericRecordArray(String fieldName, GenericRecord[] value) {
+        return write(fieldName, value, FieldType.PORTABLE_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeByteArray(String fieldName, byte[] value) {
+        return write(fieldName, value, FieldType.BYTE_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeBooleanArray(String fieldName, boolean[] value) {
+        return write(fieldName, value, FieldType.BOOLEAN_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeCharArray(String fieldName, char[] value) {
+        return write(fieldName, value, FieldType.CHAR_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeIntArray(String fieldName, int[] value) {
+        return write(fieldName, value, FieldType.INT_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeLongArray(String fieldName, long[] value) {
+        return write(fieldName, value, FieldType.LONG_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeDoubleArray(String fieldName, double[] value) {
+        return write(fieldName, value, FieldType.DOUBLE_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeFloatArray(String fieldName, float[] value) {
+        return write(fieldName, value, FieldType.FLOAT_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeShortArray(String fieldName, short[] value) {
+        return write(fieldName, value, FieldType.SHORT_ARRAY);
+    }
+
+    @Override
+    public GenericRecord.Builder writeUTFArray(String fieldName, String[] value) {
+        return write(fieldName, value, FieldType.UTF_ARRAY);
+    }
+
+    private GenericRecord.Builder write(String fieldName, Object value, FieldType fieldType) {
+        FieldDefinition fd = check(fieldName, fieldType);
+        int index = fd.getIndex();
+        if (objects[index] != null) {
+            if (!isClone) {
+                throw new HazelcastSerializationException("It is illegal to the overwrite the field");
+            } else if (isOverWritten[index]) {
+                throw new HazelcastSerializationException("Field can only overwritten once with `cloneWithBuilder`");
```

Comment:
Because it will be written directly to the byte-array in the new format. And if I allow that, it might be complex to implement.

Reply:
yes, it will be written into byte-array eventually but here you only keep an array of objects, hence it should not harm to overwrite the object at the same index until you really serialize into byte array, right?

Reply:
I am thinking about the new format and there we will write to the byte-array early in order not to create temporary lists/maps. And note that the contract comes from the interface. The behaviour does not just belong to `PortableGenericRecordBuilder` . It is the contract of the `Builder`